### DISCUSSION TITLE
Gift Card Transaction Mutations

### DISCUSSION THREAD

**Discussion by JakubKor** [saleor/saleor#12011] [Ideas]
### What I'm trying to achieve
I want to be able to create gift card transactions in an external app without using an order to do so.

### Solution for now
For now, it's only possible to use giftCardUpdate mutation to update gift card limits and implement rollbacks.

### Proposed solution
* Implement addGiftCardTransaction mutation that will handle adding transactions in gift cards.
e.g. addGiftCardTransaction(id, amount){ eventId: ID! giftCard: GiftCard errors: [GiftCardTransactionErrors!]! }

* Implement rollbackGiftCardTransaction mutation that will handle transaction rollbacks in gift cards.
e.g. rollbackGiftCardTransaction(id, eventId) { giftCard: GiftCard errors: [GiftCardRollbackTransactionErrors!]! }

### Example Usage
An app that handles local stores that use gift cards bought in the web store.

**Comment by timuric** [comment]
Hi @JakubKor could you please describe your business use case?

**Reply by JakubKor** [reply]
Hi @timuric thanks for your response

We have printed gift cards and digital gift cards they are both generated by saleor.
We would like customers to use them in local stores as well as in the webstore.
Our app communicates with local store registers to send info about gift cards(balance, possibility of usage, transactions related to this gift card). So when a Customer uses a gift card in a local store a gift card reference is sent to our app so we can register transactions on this card and send info back about it(success/failure).
Using giftCardUpdate mutation to update balance is potentially dangerous as it can introduce race conditions when the card is used in both places at once.
addGiftCardTransaction mutation will handle checking the balance and if it's possible subtract the balance by the provided amount and return an error when it's not possible to subtract.
